lib/sinon/spy.js

```diff
@@ -170,25 +158,30 @@ var spyApi = {
     },
 
     invoke: function invoke(func, thisValue, args) {
-        var matching = matchingFake(this.fakes, args);
-        var exception, returnValue;
+        var matchings = this.matchingFakes(args);
+        var currentCallId = callId++;
+        var exception, returnValue, i;
 
         incrementCallCount.call(this);
         push.call(this.thisValues, thisValue);
         push.call(this.args, args);
-        push.call(this.callIds, callId++);
+        push.call(this.callIds, currentCallId);
+        if (matchings) {
```

Comment:
If `matchingFakes` always returns arrays, then you can simplify this code (Sinon.JS codebase uses ES5.1, so the array extras are allowed).

```javascript
matchings.forEach(function (matching) {
    incrementCallCount.call(matchings);
    push.call(matchings.thisValues, thisValue);
    push.call(matchings.args, args);
    push.call(matchings.callIds, currentCallId);
});
```